File: BaseMedia/src/main/java/com/ashlikun/media/video/play/EasyVideoSystem.java
package com.ashlikun.media.video.play;

import android.media.AudioManager;
import android.media.MediaPlayer;
import android.text.TextUtils;
import android.view.Surface;
import android.widget.Toast;

import com.ashlikun.media.R;
import com.ashlikun.media.video.EasyMediaInterface;
import com.ashlikun.media.video.EasyMediaManager;

import java.lang.reflect.Method;
import java.util.Map;

/**
 * 作者　　: 李坤
 * 创建时间: 2017/12/01 16:25
 * 邮箱　　：496546144@qq.com
 * <p>
 * 功能介绍：实现系统的播放引擎
 */

public class EasyVideoSystem extends EasyMediaInterface
        implements MediaPlayer.OnPreparedListener,
        MediaPlayer.OnCompletionListener,
        MediaPlayer.OnBufferingUpdateListener,
        MediaPlayer.OnErrorListener,
        MediaPlayer.OnInfoListener,
        MediaPlayer.OnSeekCompleteListener,
        MediaPlayer.OnVideoSizeChangedListener {
    /**
     * 是否调用过暂定，调用后在准备好的时候不能直接播放
     */
    private boolean isPreparedPause = false;
    public MediaPlayer mediaPlayer;

    @Override
    public void setPreparedPause(boolean preparedPause) {
        isPreparedPause = preparedPause;
    }

    @Override
    public void start() {
        EasyMediaManager.pauseOther(easyMediaManager);
        if (mediaPlayer != null) {
            mediaPlayer.start();
        }
    }

    @Override
    public void stop() {
        if (mediaPlayer != null) {
            mediaPlayer.stop();
        }
    }

    @Override
    public void prepare() {
        try {
            if (isPlaying()) {
                stop();
            }
            release();
            mediaPlayer = new MediaPlayer();
            if (getCurrentDataSource() == null) {
                Toast.makeText(context, context.getText(R.string.easy_video_no_url), Toast.LENGTH_SHORT).show();
                onError(mediaPlayer, -2, -2);
                return;
            }
            mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
            //  mediaPlayer.setLooping((boolean) dataSource[1]);
            mediaPlayer.setOnPreparedListener(EasyVideoSystem.this);
            mediaPlayer.setOnCompletionListener(EasyVideoSystem.this);
            mediaPlayer.setOnBufferingUpdateListener(EasyVideoSystem.this);
            mediaPlayer.setScreenOnWhilePlaying(true);
            mediaPlayer.setOnErrorListener(EasyVideoSystem.this);
            mediaPlayer.setOnInfoListener(EasyVideoSystem.this);
            mediaPlayer.setOnVideoSizeChangedListener(EasyVideoSystem.this);
            mediaPlayer.setOnSeekCompleteListener(this);
            mediaPlayer.setLooping(getCurrentDataSource().isLooping());
            if (!TextUtils.isEmpty(getCurrentDataSource().getUrl())) {
                if (getCurrentDataSource().getHeaders() != null) {
                    Class<MediaPlayer> clazz = MediaPlayer.class;
                    Method method = clazz.getDeclaredMethod("setDataSource", String.class, Map.class);
                    method.invoke(mediaPlayer, getCurrentDataSource().getUrl(), getCurrentDataSource().getHeaders());
                } else {
                    mediaPlayer.setDataSource(getCurrentDataSource().getUrl());
                }
            } else if (getCurrentDataSource().getUri() != null && !TextUtils.isEmpty(getCurrentDataSource().getUri().toString())) {
                if (getCurrentDataSource().getHeaders() != null) {
                    mediaPlayer.setDataSource(context, getCurrentDataSource().getUri(), getCurrentDataSource().getHeaders());
                } else {
                    mediaPlayer.setDataSource(context, getCurrentDataSource().getUri());
                }
            } else if (getCurrentDataSource().getFileDescriptor() != null) {
                mediaPlayer.setDataSource(getCurrentDataSource().getFileDescriptor());
            } else {
                Toast.makeText(context, context.getText(R.string.easy_video_no_url), Toast.LENGTH_SHORT).show();
                onError(mediaPlayer, -2, -2);
                return;
            }
            mediaPlayer.prepareAsync();
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(context, context.getText(R.string.easy_video_no_url), Toast.LENGTH_SHORT).show();
            onError(mediaPlayer, -2, -2);
        }
    }

    @Override
    public void pause() {
        if (mediaPlayer != null) {
            mediaPlayer.pause();
        }
    }

    @Override
    public boolean isPlaying() {
        if (mediaPlayer == null) {
            return false;
        }
        if (mediaPlayer != null) {
            return mediaPlayer.isPlaying();
        }
        return false;
    }

    @Override
    public void seekTo(long time) {
        if (mediaPlayer == null) {
            return;
        }
        mediaPlayer.seekTo((int) time);
    }

    @Override
    public void release() {
        if (mediaPlayer != null) {
            mediaPlayer.release();
            mediaPlayer = null;
        }
    }

    @Override
    public long getCurrentPosition() {
        if (mediaPlayer == null) {
            return 0;
        }
        return mediaPlayer.getCurrentPosition();
    }

    @Override
    public int getBufferedPercentage() {
        return -1;
    }

    @Override
    public long getDuration() {
        if (mediaPlayer == null) {
            return 0;
        }
        return mediaPlayer.getDuration();
    }

    @Override
    public void setSurface(Surface surface) {
        if (mediaPlayer == null) {
            return;
        }
        mediaPlayer.setSurface(surface);
    }

    @Override
    public void onPrepared(MediaPlayer mediaPlayer) {
        if (mediaPlayer == null) {
            return;
        }
        if (!isPreparedPause) {
            mediaPlayer.start();
            if (easyMediaManager.getMediaPlay() == this)
                easyMediaManager.getHandlePlayEvent().onPrepared();
        } else {
            isPreparedPause = false;
        }
    }

    @Override
    public void onCompletion(MediaPlayer mediaPlayer) {
        if (easyMediaManager.getMediaPlay() == this)
            easyMediaManager.getHandlePlayEvent().onCompletion(this);
    }

    @Override
    public void onBufferingUpdate(MediaPlayer mediaPlayer, final int percent) {
        if (easyMediaManager.getMediaPlay() == this)
            easyMediaManager.getHandlePlayEvent().setBufferProgress(percent);
    }

    @Override
    public boolean onError(MediaPlayer mediaPlayer, final int what, final int extra) {
        if (easyMediaManager.getMediaPlay() == this)
            easyMediaManager.getHandlePlayEvent().onError(what, extra);
        return true;
    }

    @Override
    public boolean onInfo(MediaPlayer mediaPlayer, final int what, final int extra) {
        if (easyMediaManager.getMediaPlay() == this)
            easyMediaManager.getHandlePlayEvent().onInfo(what, extra);
        return false;
    }

    @Override
    public void onVideoSizeChanged(MediaPlayer mediaPlayer, int width, int height) {
        if (easyMediaManager.getMediaPlay() == this)
            easyMediaManager.getHandlePlayEvent().onVideoSizeChanged(width, height);
    }

    @Override
    public void onSeekComplete(MediaPlayer mp) {
        if (easyMediaManager.getMediaPlay() == this)
            easyMediaManager.getHandlePlayEvent().onSeekComplete();
    }
}

File: BaseMedia/src/main/java/com/ashlikun/media/video/play/EasyVideoSystem.kt
package com.ashlikun.media.video.play

import android.media.AudioManager
import android.media.MediaPlayer
import android.text.TextUtils
import android.view.Surface
import android.widget.Toast
import com.ashlikun.media.R
import com.ashlikun.media.video.EasyMediaInterface
import com.ashlikun.media.video.EasyMediaManager
import com.ashlikun.media.video.EasyMediaManager.Companion.pauseOther
import com.ashlikun.media.video.VideoUtils

/**
 * 作者　　: 李坤
 * 创建时间: 2017/12/01 16:25
 * 邮箱　　：496546144@qq.com
 *
 *
 * 功能介绍：实现系统的播放引擎
 */
class EasyVideoSystem(override val manager: EasyMediaManager) : EasyMediaInterface(manager), MediaPlayer.OnPreparedListener,
    MediaPlayer.OnCompletionListener, MediaPlayer.OnBufferingUpdateListener,
    MediaPlayer.OnErrorListener, MediaPlayer.OnInfoListener, MediaPlayer.OnSeekCompleteListener, MediaPlayer.OnVideoSizeChangedListener {
    /**
     * 是否调用过暂定，调用后在准备好的时候不能直接播放
     */
    private var isPreparedPause = false
    var mediaPlayer: MediaPlayer? = null
    override fun setPreparedPause(preparedPause: Boolean) {
        isPreparedPause = preparedPause
    }

    override fun start() {
        pauseOther(manager)
        if (mediaPlayer != null) {
            mediaPlayer!!.start()
        }
    }

    override fun stop() {
        if (mediaPlayer != null) {
            mediaPlayer!!.stop()
        }
    }

    override fun prepare() {
        try {
            if (isPlaying) {
                stop()
            }
            release()
            mediaPlayer = MediaPlayer()
            if (currentDataSource == null) {
                Toast.makeText(VideoUtils.context, VideoUtils.context.getText(R.string.easy_video_no_url), Toast.LENGTH_SHORT).show()
                onError(mediaPlayer!!, -2, -2)
                return
            }
            mediaPlayer!!.setAudioStreamType(AudioManager.STREAM_MUSIC)
            //  mediaPlayer.setLooping((boolean) dataSource[1]);
            mediaPlayer!!.setOnPreparedListener(this@EasyVideoSystem)
            mediaPlayer!!.setOnCompletionListener(this@EasyVideoSystem)
            mediaPlayer!!.setOnBufferingUpdateListener(this@EasyVideoSystem)
            mediaPlayer!!.setScreenOnWhilePlaying(true)
            mediaPlayer!!.setOnErrorListener(this@EasyVideoSystem)
            mediaPlayer!!.setOnInfoListener(this@EasyVideoSystem)
            mediaPlayer!!.setOnVideoSizeChangedListener(this@EasyVideoSystem)
            mediaPlayer!!.setOnSeekCompleteListener(this)
            mediaPlayer!!.isLooping = currentDataSource!!.isLooping
            if (!TextUtils.isEmpty(currentDataSource!!.url)) {
                if (currentDataSource!!.headers != null) {
                    val clazz = MediaPlayer::class.java
                    val method = clazz.getDeclaredMethod("setDataSource", String::class.java, MutableMap::class.java)
                    method.invoke(mediaPlayer, currentDataSource!!.url, currentDataSource!!.headers)
                } else {
                    mediaPlayer!!.setDataSource(currentDataSource!!.url)
                }
            } else if (currentDataSource!!.uri != null && !TextUtils.isEmpty(currentDataSource!!.uri.toString())) {
                if (currentDataSource!!.headers != null) {
                    mediaPlayer!!.setDataSource(VideoUtils.context, currentDataSource!!.uri!!, currentDataSource!!.headers)
                } else {
                    mediaPlayer!!.setDataSource(VideoUtils.context, currentDataSource!!.uri!!)
                }
            } else if (currentDataSource!!.fileDescriptor != null) {
                mediaPlayer!!.setDataSource(currentDataSource!!.fileDescriptor)
            } else {
                Toast.makeText(VideoUtils.context, VideoUtils.context.getText(R.string.easy_video_no_url), Toast.LENGTH_SHORT).show()
                onError(mediaPlayer!!, -2, -2)
                return
            }
            mediaPlayer!!.prepareAsync()
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(VideoUtils.context, VideoUtils.context.getText(R.string.easy_video_no_url), Toast.LENGTH_SHORT).show()
            onError(mediaPlayer!!, -2, -2)
        }
    }

    override fun pause() {
        if (mediaPlayer != null) {
            mediaPlayer!!.pause()
        }
    }

    override val isPlaying: Boolean
        get() {
            if (mediaPlayer == null) {
                return false
            }
            return if (mediaPlayer != null) {
                mediaPlayer!!.isPlaying
            } else false
        }

    override fun seekTo(time: Long) {
        mediaPlayer?.seekTo(time.toInt())
    }

    override fun release() {
        mediaPlayer?.release()
        mediaPlayer = null
    }

    override val currentPosition: Long
        get() = mediaPlayer?.currentPosition?.toLong() ?: 0
    override val bufferedPercentage: Int
        get() = -1
    override val duration: Long
        get() = mediaPlayer?.duration?.toLong() ?: 0

    override fun setSurface(surface: Surface) {
        mediaPlayer?.setSurface(surface)
    }

    override fun onPrepared(mediaPlayer: MediaPlayer) {
        if (!isPreparedPause) {
            mediaPlayer.start()
            if (manager.mediaPlay === this) manager.handlePlayEvent.onPrepared()
        } else {
            isPreparedPause = false
        }
    }

    override fun onCompletion(mediaPlayer: MediaPlayer) {
        if (manager.mediaPlay === this) manager.handlePlayEvent.onCompletion(this)
    }

    override fun onBufferingUpdate(mediaPlayer: MediaPlayer, percent: Int) {
        if (manager.mediaPlay === this) manager.handlePlayEvent.setBufferProgress(percent)
    }

    override fun onError(mediaPlayer: MediaPlayer, what: Int, extra: Int): Boolean {
        if (manager.mediaPlay === this) manager.handlePlayEvent.onError(what, extra)
        return true
    }

    override fun onInfo(mediaPlayer: MediaPlayer, what: Int, extra: Int): Boolean {
        if (manager.mediaPlay === this) manager.handlePlayEvent.onInfo(what, extra)
        return false
    }

    override fun onVideoSizeChanged(mediaPlayer: MediaPlayer, width: Int, height: Int) {
        if (manager.mediaPlay === this) manager.handlePlayEvent.onVideoSizeChanged(width, height)
    }

    override fun onSeekComplete(mp: MediaPlayer) {
        if (manager.mediaPlay === this) manager.handlePlayEvent.onSeekComplete()
    }
}